File: algorithms/0102-binary-tree-level-order-traversal/Solution.kt
// Binary Tree Level Order Traversal - size-based BFS with a node queue.
// Time O(n), Space O(n).

class TreeNode(var `val`: Int) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

class Solution {
    fun levelOrder(root: TreeNode?): List<List<Int>> {
        val result = mutableListOf<List<Int>>()
        if (root == null) return result
        val q = ArrayDeque<TreeNode>()
        q.addLast(root)
        while (q.isNotEmpty()) {
            val size = q.size
            val level = mutableListOf<Int>()
            repeat(size) {
                val node = q.removeFirst()
                level.add(node.`val`)
                node.left?.let { q.addLast(it) }
                node.right?.let { q.addLast(it) }
            }
            result.add(level)
        }
        return result
    }
}

File: algorithms/0102-binary-tree-level-order-traversal/solution.py
# Binary Tree Level Order Traversal - size-based BFS with a node queue.
# Time O(n), Space O(n).

from collections import deque


class TreeNode:
    def __init__(self, val: int = 0, left: "TreeNode | None" = None, right: "TreeNode | None" = None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: TreeNode | None) -> list[list[int]]:
        result: list[list[int]] = []
        if root is None:
            return result
        q: deque[TreeNode] = deque([root])
        while q:
            size = len(q)
            level: list[int] = []
            for _ in range(size):
                node = q.popleft()
                level.append(node.val)
                if node.left is not None:
                    q.append(node.left)
                if node.right is not None:
                    q.append(node.right)
            result.append(level)
        return result
